src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs

```diff
@@ -99,6 +100,384 @@ public void Move(int oldIndex, int newIndex)
             MoveItem(oldIndex, newIndex);
         }
 
+
+        /// <summary>
+        /// Adds the elements of the specified collection to the end of the <see cref="ObservableCollection{T}"/>.
+        /// </summary>
+        /// <param name="collection">
+        /// The collection whose elements should be added to the end of the <see cref="ObservableCollection{T}"/>.
+        /// The collection itself cannot be null, but it can contain elements that are null, if type T is a reference type.
+        /// </param>
+        /// <exception cref="ArgumentNullException"><paramref name="collection"/> is null.</exception>
+        public void AddRange(IEnumerable<T> collection)
+        {
+            InsertRange(Count, collection);
+        }
+
+        /// <summary>
+        /// Inserts the elements of a collection into the <see cref="ObservableCollection{T}"/> at the specified index.
+        /// </summary>
+        /// <param name="index">The zero-based index at which the new elements should be inserted.</param>
+        /// <param name="collection">The collection whose elements should be inserted into the List<T>.
+        /// The collection itself cannot be null, but it can contain elements that are null, if type T is a reference type.</param>                
+        /// <exception cref="ArgumentNullException"><paramref name="collection"/> is null.</exception>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="index"/> is not in the collection range.</exception>
+        public void InsertRange(int index, IEnumerable<T> collection)
+        {
+            if (collection == null)
+                throw new ArgumentNullException(nameof(collection));
+
+            if (index > Count)
+                throw new ArgumentOutOfRangeException(nameof(index));
+
+            if (collection is ICollection<T> list)
+            {
+                if (list.Count == 0)
+                {
+                    return;
+                }
+                if (!(list is IList))
+                {
+                    list = new List<T>(list);
+                }
+            }
+            else if (!ContainsAny(collection))
+            {
+                return;
+            }
+            else
+            {
+                list = new List<T>(collection);
+            }
+
+            if (list.Count == 1)
+            {
+                InsertItem(index, ((IList<T>)list)[0]);
```

Comment:
It will throw if `collection`  implements `IEnumerable<T>,  ICollection<T>, IList` but not `IList<T>`

Reply:
Will fix.

Reply:
We're starting with `IEnumerable<T>`, I think it's a very low possibility for a class to implement the first two generic while having only `IList` as non-generic.
I changed the code in the class by rewriting it in a different manner (will commit later), but should we encounter a similar situation, IMHO we should just add `Debug.Assert(collection is IList<T>)`.

Reply:
We're starting with `IEnumerable<T>`, I think it's a very low possibility for a class to implement the first two generic while having only `IList` as non-generic.
I changed the code in the class by rewriting it in a different manner (will commit later), but should we encounter a similar situation, IMHO we should just add `Debug.Assert(collection is IList<T>)`, or an explicit check with `NotSupportedException`, unless we can come up with concrete examples.

Reply:
BTW, #23578 targets part of the issue of having to check between the several types of collections where the generic and non-generic ones aren't related.

Reply:
Fixed.

Reply:
The contract says it only need to be `IEnumerable<T>`. The method should work correctly for any  correct enumerable.
Implementing additional interfaces is just an optimization opportunity.

Reply:
This is now been fixed anyway.